be/src/connector/binlog_connector.cpp

```diff
@@ -0,0 +1,212 @@
+            meta_fields.emplace_back(binlog_meta_map[BINLOG_SEQ_ID]);
+        } else if (slot->col_name() == BINLOG_TIMESTAMP) {
+            meta_column_slot_index.push_back(slot_index);
+            meta_fields.emplace_back(binlog_meta_map[BINLOG_TIMESTAMP]);
+        } else {
+            std::stringstream ss;
+            ss << "invalid field name: " << slot->col_name();
+            LOG(WARNING) << ss.str();
+            return Status::InternalError(ss.str());
+        }
+    }
+
+    if (data_column_cids.empty()) {
+        return Status::InternalError("failed to build binlog schema, no materialized data slot!");
+    }
+
+    const TabletSchema& tablet_schema = _tablet->tablet_schema();
+    VectorizedSchema schema = ChunkHelper::convert_schema_to_format_v2(tablet_schema, data_column_cids);
+    for (int32_t i = 0; i < meta_column_slot_index.size(); i++) {
+        uint32_t index = meta_column_slot_index[i];
+        if (index >= schema.num_fields()) {
+            schema.append(meta_fields[i]);
+        } else {
+            schema.insert(index, meta_fields[i]);
+        }
+    }
+
+    return schema;
+}
+
+Status BinlogDataSource::_mock_chunk(Chunk* chunk) {
+    int32_t num = _chunk_num.fetch_add(1, std::memory_order_relaxed);
+    if (num >= 2) {
+        return Status::EndOfFile(fmt::format("Has sent {} chunks", num));
+    }
+
+    for (int row = 0; row < 10; row++) {
+        for (int col = 0; col < chunk->num_columns(); col++) {
+            LogicalType type = _binlog_read_schema.field(col)->type()->type();
+            Datum datum;
+            switch (type) {
+            case TYPE_TINYINT: {
+                int8_t val = row + col;
+                datum = Datum(val);
+                break;
+            }
+            case TYPE_SMALLINT: {
+                int16_t val = row + col;
+                datum = Datum(val);
+                break;
+            }
+            case TYPE_INT: {
+                int32_t val = row + col;
+                datum = Datum(val);
+                break;
+            }
+            case TYPE_BIGINT: {
+                int64_t val = row + col;
+                datum = Datum(val);
+                break;
+            }
+            case TYPE_LARGEINT: {
+                int128_t val = row + col;
+                datum = Datum(val);
+                break;
+            }
+            default:
+                return Status::InternalError(fmt::format("Not support to mock type {}", type));
+            }
+            chunk->get_column_by_index(col)->append_datum(datum);
+        }
+    }
+    return Status::OK();
+}
+
+StatusOr<TabletSharedPtr> BinlogDataSource::_get_tablet() {
+    TTabletId tablet_id = _scan_range.tablet_id;
+    std::string err;
+    TabletSharedPtr tablet = StorageEngine::instance()->tablet_manager()->get_tablet(tablet_id, true, &err);
+    if (!tablet) {
+        std::stringstream ss;
+        ss << "failed to get tablet. tablet_id=" << tablet_id << ", reason=" << err;
```

Comment:
Use `fmt::format` instead of `stringstream`?

Reply:
Done